api/API_VERSIONING.md

```diff
@@ -0,0 +1,203 @@
+  feature in both locations.
+* We encourage Envoy users to move to the current stable major version from the previous one to
+  consume new functionality.
+
+As a pragmatic concession, we allow API feature additions to the previous stable major version for a
+single quarter following a major API version increment. Any changes to the previous stable major
+version must be manually reflected in a consistent manner in the current stable major version as
+well.
+
+# How to make a breaking change across major versions
+
+We maintain [backwards compatibility](#backwards-compatibility) within a major version but allow
+breaking changes across major versions. This enables API deprecations, cleanups, refactoring and
+reorganization. The Envoy APIs have a stylized workflow for achieving this. There are two prescribed
+methods, depending on whether the change is mechanical or manual.
+
+## Mechanical breaking changes
+
+Field deprecations, renames, etc. are mechanical changes that will be supported by the
+[protoxform](https://github.com/envoyproxy/envoy/tree/master/tools/protoxform) tool. These are
+guided by annotations in protobuf.
+* Deprecations are specified with the built-in protobuf deprecated option set on a message, enum,
+  field or enum value. No field may be marked as deprecated unless a replacement for this
+  functionality exists and the corresponding Envoy implementation is production ready.
+
+* Renames are specified with a `[#rename-at-next-major-version: <new name>]` protobuf comment
+  annotation.
+
+* We anticipate that `protoxform` will also support `oneof` promotion, package movement, etc. via
+  similar annotations.
+
+## Manual breaking changes
+
+A manual breaking change is distinct from the mechanical changes such as field deprecation, since in
+general it requires new code and tests to be implemented in Envoy by hand. For example, if a developer
+wants to unify `HeaderMatcher` with `StringMatcher` in the route configuration, this is a likely
+candidate for this class of change. The following steps are required:
+1. The new version of the feature, e.g. the `NewHeaderMatcher` message should be added, together
+   with referencing fields, in the current stable major version for the route configuration proto.
+2. The Envoy implementation should be changed to consume configuration from the fields added in (1).
+   Translation code (and tests) should be written to map from the existing field and messages to
+   (1).
+3. The old message/enum/field/enum value should be annotated as deprecated.
+4. At the next major version, `protoxform` will remove the deprecated version automatically.
+
+This approach ensures that API major version releases are predictable and mechanical, and has the
+bulk of the Envoy code and test changes owned by feature developers, rather than the API owners.
+There will be no major `vN` initiative to address technical debt beyond that enabled by the above
+process.
+
+# One Definition Rule (ODR)
```

Comment:
If the major version is always part of the package name, doesn't that prevent ODR violations?  I don't quite understand the problem here -- can you give a concrete example to illustrate?

Reply:
`envoy.foo.v4` refers to `envoy.bar.v4` and `envoy.baz.v3`. `envoy.bar.v4` refers to `envoy.blah.v2` while `envoy.baz.v3` refers to `envoy.blah.v3`. Voila, transitively `envoy.foo.v4` references `envoy.blah.v2` and `envoy.blah.v3`.

Reply:
Sure, but since `envoy.blah.v2` and `envoy.blah.v3` are different namespaces, how does that cause an ODR violation?

Reply:
It doesn't at a C++ level, but it does at a higher level, since by allowing these multiple definitions, `envoy.blah` now needs to maintain an arbitrary number of stable versions, rather than just the trailing two versions. That includes all the code backing these versions. By forcing an "ODR-style" rule here, we avoid this quagmire.